pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+                                                    array_fmt_bytes)
+
+        return DmapArray(array_name, array_value, array_type,
+                         array_dimension, array_shape)
+
+    def read_data(self, data_type_fmt, data_fmt_bytes: int):
+        """
+        Reads in individual data type from the byte array
+        Given a dmap data type.
+
+        Parameter
+        ---------
+        data_type : int
+            dmap data type numerical value for knowing how many bytes to read.
+
+        Return
+        ------
+        data : data_type
+            returns the data that was store in the byte array.
+            Has the data type of the dmap data type passed in.
+
+        Raises
+        ------
+        CursorError
+            if cursor is not correctly aligned or at the correct position.
+        DmapDataError
+            if trying to read a DMAP data type
+
+        See Also
+        --------
+        Struct.unpack : converts the bytes array values from a C structure to
+        python objects
+
+        See DEVELOPER_README.md for more information on DMAP data types and
+        Struct.unpack.
+        """
+        if self.cursor >= self.dmap_end_bytes:
+            message = "Error: Cursor {cursor} extends"\
+                    " out of buffer {end_bytes}."\
+                    "".format(cursor=self.cursor,
+                              end_bytes=self.dmap_end_bytes)
+            raise pydmap_exceptions.CursorError(self.cursor, message=message)
+
+        cursor_offset = self.dmap_end_bytes - self.cursor
+        if cursor_offset < data_fmt_bytes:
+            message = "Error: Byte offsets {offset} into buffer {fmt}"\
+                    " are not aligned.".format(offset=cursor_offset,
+                                               fmt=data_fmt_bytes)
+            raise pydmap_exceptions.CursorError(self.cursor, message=message)
+
+        # struct.unpack is a python method the converts C struct as python bytes
+        # to python values. See DEVELOPER_README.md for more information.
+        # TODO: how to handle this gracefully? Return None and check later than
+        #       continue in the loop?
+        if data_type_fmt == 'c':
+            data = self.dmap_bytearr[self.cursor]
+            self.cursor += data_fmt_bytes
+            return data
+        elif data_type_fmt == 's':
+            byte_counter = 0
+            while self.dmap_bytearr[self.cursor + byte_counter] != 0 or self.cursor + byte_counter >= self.dmap_end_bytes:
+                # add 1 byte to byte_counter because a string is a list of chars
+                byte_counter += 1  # 1 is the number of bytes of a char
+            self.bytes_check(self.cursor+byte_counter, 'cursor', self.dmap_end_bytes, 'Total number of Bytes')
+
+            # format byte counter to a string format for unpacking
+            # WARNING: not using format for performance purposes
+            char_count = '%ss'%byte_counter
```

Comment:
change % to .format

Reply:
Read the comment above the line. I use `%` because of a performance reason. 
`.format` is slower than `%` and since this method gets called on every parameter name the ms add up.